### PR TITLE
FEATURE: Allow email image embed with secure media

### PR DESCRIPTION
**Pull request by martin-brennan**
This PR introduces a few important changes to secure media redaction in emails. First of all, two new site settings have been introduced:

* `secure_media_allow_embed_images_in_emails`: If enabled we will embed secure images in emails instead of redacting them.
* `secure_media_max_email_embed_image_size_kb`: The cap to the size of the secure image we will embed, defaulting to 1mb, so the email does not become too big. Max is 10mb. Works in tandem with `email_total_attachment_size_limit_kb`.

`Email::Sender` will now attach images to the email based on these settings. The sender will also call `inline_secure_images` in `Email::Styles` after secure media is redacted and attachments are added to replace redaction messages with attached images. I went with attachment and `cid` URLs because base64 image support is _still_ flaky in email clients.

All redaction of secure media is now handled in `Email::Styles` and calls out to `PrettyText.strip_secure_media` to do the actual stripping and replacing with placeholders. `app/mailers/group_smtp_mailer.rb` and `app/mailers/user_notifications.rb` no longer do any stripping because they are earlier in the pipeline than `Email::Styles`.

Finally the redaction notice has been restyled and includes a link to the media that the user can click, which will show it to them if they have the necessary permissions.

image

### CODE REVIEW

File: lib/email/styles.rb [hunk 1]
```diff
@@ -237,10 +236,40 @@ def plugin_styles
       @@plugin_callbacks.each { |block| block.call(@fragment, @opts) }
     end
 
+    def inline_secure_images(attachments)
```
Comment (martin-brennan) [inline lib/email/styles.rb:239]: I feel this method is a little ugly with lots of data munging, suggestions to clean up would be welcome.

File: spec/components/email/sender_spec.rb [hunk 2]
```diff
@@ -404,6 +404,72 @@
...
+      end
+    end
+
+    it "adds only non-image uploads as attachments to the email" do
+      SiteSetting.email_total_attachment_size_limit_kb = 10_000
+      Email::Sender.new(message, :valid_type).send
+
+      expect(message.attachments.length).to eq(3)
+      expect(message.attachments.map(&:filename))
+        .to contain_exactly(*[small_pdf, large_pdf, csv_file].map(&:original_filename))
```
Comment (pmusaraj) [inline spec/components/email/sender_spec.rb:470]: Would it makes sense to have the spec check that the non-secure image here is in the markup?

File: spec/models/upload_spec.rb [hunk 3]
```diff
@@ -434,41 +434,6 @@ def enable_secure_media
     end
   end
 
-  context ".signed_url_from_secure_media_url" do
```
Comment (pmusaraj) [inline spec/models/upload_spec.rb:437]: It's not clear to me why these specs have been removed.

**Review by pmusaraj — APPROVED** [review]
Overall this looks great, just the few minor comments on the specs.

File: spec/models/upload_spec.rb (hunk 3 above)
Comment (martin-brennan) [reply spec/models/upload_spec.rb:437]: That makes 2 of us! Looks like it was just some merge shenanigans from those other secure media fixes I made, fixed now.

File: spec/components/email/sender_spec.rb (hunk 2 above)
Comment (martin-brennan) [reply spec/components/email/sender_spec.rb:470]: @pmusaraj yes it totally would, I will add that spec.

**Pull request merged**